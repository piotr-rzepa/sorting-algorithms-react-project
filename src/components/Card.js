/*
 * Komponent bezklasowy będący "kartą" -> Blokiem z informacją np. o  szybkości działania
 */

/* eslint-disable react/prop-types */
import React from 'react';

export default function Card({ title, subtitle }) {
	return (
		<div className="Card">
			<h2>{title}</h2>
			<p>{subtitle}</p>
		</div>
	);
}
