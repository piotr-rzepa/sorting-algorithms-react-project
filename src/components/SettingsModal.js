/*
 * Zewnętrzny modal pozwalający użytkownikowi na konfigurację opcji do przeprowadzania testów
 */

/* eslint-disable react/prop-types */
import React, { useContext } from 'react';
import Modal from 'react-modal';
import { useDispatch, useSelector } from 'react-redux';
import {
	setSizeDataset,
	setSizeArray,
	setDataOrder,
	setSortingOrder,
	setArrayGain
} from '../actions/tests';
import { ThemeContext } from './ThemeContext';
import { styleModal, styleButtons, styleSearchBar } from '../styles/styles';

export default function SettingsModal({
	isOpen,
	onRequestClose,
	contentLabel
}) {
	const {
		datasetSize,
		arrayStartSize,
		dataOrder,
		sortingOrder,
		arrayEndSize
	} = useSelector((state) => state.tests);
	const dispatch = useDispatch();
	const darkTheme = useContext(ThemeContext);
	const [currentModalStyle, currentButtonStyle, currentInputStyle] = darkTheme
		? [styleModal.darkStyle, styleButtons.darkStyle, styleSearchBar.darkStyle]
		: [styleModal.lightStyle, styleButtons.lightStyle, styleSearchBar.lightStyle];
	return (
		<Modal
			isOpen={isOpen}
			onRequestClose={onRequestClose}
			contentLabel={contentLabel}
			closeTimeoutMS={200}
			className="modal"
			style={{ overlay: { background: currentModalStyle.backgroundColor } }}
		>
			<h2>Algorithm dataset settings</h2>
			<div className="numeric-settings">
				<div className="settings-title">
					<label htmlFor="dataset-length" name="dataset-length">
						Number of tests per array length:
					</label>
					<label htmlFor="starting-array-length" name="starting-array-length">
						Star array length:
					</label>
					<label htmlFor="array-gain" name="array-gain">
						End array length:
					</label>
				</div>
				<div className="settings-input">
					<input
						type="number"
						min="10"
						max="10000"
						step="10"
						value={datasetSize}
						onChange={(e) => dispatch(setSizeDataset(parseInt(e.target.value, 10)))}
						style={currentInputStyle}
						required
					></input>
					<input
						type="number"
						min="10"
						max="1000"
						step="10"
						value={arrayStartSize}
						onChange={(e) => dispatch(setSizeArray(parseInt(e.target.value, 10)))}
						style={currentInputStyle}
						required
					></input>

					<input
						type="number"
						min="100"
						max="10000"
						step="10"
						value={arrayEndSize}
						onChange={(e) => dispatch(setArrayGain(parseInt(e.target.value, 10)))}
						style={currentInputStyle}
						required
					></input>
				</div>
			</div>
			<div className="select-settings">
				<p>Select data order:</p>
				<div className="select-settings-single">
					<input
						type="radio"
						id="best"
						name="scenario"
						value="best"
						checked={dataOrder === 'best' ? true : false}
						onChange={() => dispatch(setDataOrder('best'))}
					></input>
					<label htmlFor="best"> Best scenario (data already sorted)</label>
				</div>
				<div className="select-settings-single">
					<input
						type="radio"
						id="worst"
						name="scenario"
						value="worst"
						checked={dataOrder === 'worst' ? true : false}
						onChange={() => dispatch(setDataOrder('worst'))}
					></input>
					<label htmlFor="worst"> Worst scenario (data sorted in reverse)</label>
				</div>
				<div className="select-settings-single">
					<input
						type="radio"
						id="random"
						name="scenario"
						value="random"
						checked={dataOrder === 'random' ? true : false}
						onChange={() => dispatch(setDataOrder('random'))}
					></input>
					<label htmlFor="random"> Random scenario (unfiorm distribution)</label>
				</div>
			</div>
			<div className="select-settings">
				<p>Select sorting order:</p>
				<div className="select-settings-single">
					<input
						type="radio"
						id="asc"
						name="sorting"
						value="asc"
						checked={sortingOrder === 'asc' ? true : false}
						onChange={() => dispatch(setSortingOrder('asc'))}
					></input>
					<label>Ascending order</label>
				</div>
				<div className="select-settings-single">
					<input
						type="radio"
						id="desc"
						name="sorting"
						value="desc"
						checked={sortingOrder === 'desc' ? true : false}
						onChange={() => dispatch(setSortingOrder('desc'))}
					></input>
					<label>Descending order</label>
				</div>
			</div>
			<button
				className="data-options-button--modal"
				type="button"
				onClick={() => onRequestClose()}
				style={currentButtonStyle}
			>
				Save and close
			</button>
		</Modal>
	);
}
