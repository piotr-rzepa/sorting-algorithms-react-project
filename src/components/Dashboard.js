/*
 * Komponent bezklasowy będący stroną główną aplikacji
 */

import React, { useState, useContext } from 'react';
import Card from './Card';
import ChartBar from './ChartBar';
import ChartPie from './ChartPie';
import AlgoList from './AlgoList';
import SettingsModal from './SettingsModal';
import { ReactComponent as EditSVG } from '../images/edit.svg';
import GroupSVG from '../images/group.svg';
import ClockSVG from '../images/clock.svg';
import DecreasingSVG from '../images/decreasing.svg';
import TrophySVG from '../images/trophy.svg';
import { useSelector } from 'react-redux';
import selectFastestSlowestAlgo from '../selectors/fastslowestSelector';
import numeral from 'numeral';
import { ThemeContext } from './ThemeContext';
import { styleMain, styleButtons } from '../styles/styles';

export default function Dashboard() {
	// Pobieranie motywu [jasny/ciemny]
	const darkTheme = useContext(ThemeContext);

	//Ustawianie motywu w zależności od kontekstu
	const currentStyle = darkTheme ? styleMain.darkStyle : styleMain.lightStyle;
	const numberOfAlgorithms = useSelector((state) => state.algorithms.length);
	const [fastest, slowest, average] = useSelector((state) =>
		selectFastestSlowestAlgo(state)
	);
	const [modalIsOpen, setIsOpen] = useState(false);

	return (
		<div className="Dashboard" style={currentStyle}>
			<SettingsModal
				isOpen={modalIsOpen}
				onRequestClose={() => setIsOpen(false)}
				contentLabel="Example Modal"
			/>
			<div className="overview">
				<div className="overview-title">
					<h1>Dashboard</h1>
					<p>Overview of sorting algorithms</p>
				</div>
				<button
					className="data-options-button"
					type="button"
					onClick={() => setIsOpen(() => true)}
					style={darkTheme ? styleButtons.darkStyle : styleButtons.lightStyle}
				>
					<EditSVG width="24px" height="24px" />
					Manage data options
				</button>
			</div>
			<div className="small-cards">
				<Card
					title={numberOfAlgorithms}
					subtitle={'Algorithms'}
					icon={GroupSVG}
					background={{ backgroundColor: 'rgba(113,59,219,0.05)' }}
				/>
				<Card
					title={`${numeral(average * 100).format('00:00:00')}ms`}
					subtitle={'Overall average sorting time'}
					icon={ClockSVG}
					background={{ backgroundColor: 'rgba(76,184,255,0.07)' }}
				/>
				<Card
					title={slowest}
					subtitle={'Worst average sorting time'}
					icon={DecreasingSVG}
					background={{ backgroundColor: 'rgba(255,76,97,0.05)' }}
				/>
				<Card
					title={fastest}
					subtitle={'Best average sorting time'}
					icon={TrophySVG}
					background={{ backgroundColor: 'rgba(255,184,0,0.07)' }}
				/>
			</div>
			<div className="charts">
				<ChartBar width={820} height={609} />
				<AlgoList />
				<ChartPie width={387} height={609} />
			</div>
			<p style={{ textAlign: 'right', padding: '2px 0' }}>
				Made by Chili Labs & Piotr
			</p>
		</div>
	);
}
