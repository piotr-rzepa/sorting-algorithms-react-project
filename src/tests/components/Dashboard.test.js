import * as React from 'react';
import * as reactRedux from 'react-redux';
import { shallow } from 'enzyme';
import Dashboard from '../../components/Dashboard';
import SettingsModal from '../../components/SettingsModal';
import Card from '../../components/Card';
const data = [
	{
		id: 0,
		name: 'Bubble Sort',
		description:
			'Bubble sort, sometimes referred to as sinking sort, is a simple sorting algorithm',
		method: 'Exchanging',
		sortingTime: { best: 0, worst: 0, average: 10 }
	},
	{
		id: 1,
		name: 'Quick Sort',
		description: 'Quicksort is a divide-and-conquer algorithm.',
		method: 'Partitioning',
		sortingTime: { best: 0, worst: 0, average: 100 }
	}
];

jest.mock('../../selectors/fastslowestSelector', () => {
	jest.fn().mockImplementation(() => data);
});

describe('<Dashboard />', () => {
	const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
	const useContextMock = jest.spyOn(React, 'useContext');
	const useStateMock = jest.spyOn(React, 'useState');
	const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');
	beforeEach(() => {
		useSelectorMock.mockClear();
		useContextMock.mockClear();
		useStateMock.mockClear();
		useDispatchMock.mockClear();
	});
	it('Should render component correctly', () => {
		useSelectorMock.mockImplementation(() => [0, 0, 0]);
		useContextMock.mockReturnValue(jest.fn());
		useStateMock.mockReturnValue([true, jest.fn()]);
		const wrapper = shallow(<Dashboard />);
		expect(wrapper).toMatchSnapshot();
	});

	it('Should call selectors crrectly', () => {
		useSelectorMock.mockImplementation(() => [0, 0, 0]);
		useContextMock.mockReturnValue(jest.fn());
		useStateMock.mockReturnValue([true, jest.fn()]);
		shallow(<Dashboard />);
		expect(useSelectorMock).toHaveBeenCalled();
		expect(useContextMock).toHaveBeenCalled();
		expect(useStateMock).toHaveBeenCalled();
	});

	it('Should open modal correctly', () => {
		const setStateMock = jest.fn();
		useSelectorMock.mockImplementation(() => [0, 0, 0]);
		useContextMock.mockReturnValue(jest.fn());
		useStateMock.mockReturnValue([true, setStateMock]);
		const wrapper = shallow(<Dashboard />);
		wrapper.find('button').at(0).simulate('click');
		expect(setStateMock).toHaveBeenCalled();
	});

	it('Should close modal correctly', () => {
		const setStateMock = jest.fn();
		const requestClose = jest.fn();
		useSelectorMock.mockImplementation(() => [0, 0, 0]);
		useContextMock.mockReturnValue(jest.fn());
		useStateMock.mockReturnValue([true, setStateMock]);
		useDispatchMock.mockReturnValue(jest.fn());
		const wrapper = shallow(<Dashboard />);
		wrapper.find('button').at(0).simulate('click');
		const modal = shallow(<SettingsModal onRequestClose={requestClose} />);
		modal.find('button').simulate('click');
		expect(setStateMock).toHaveBeenCalled();
		expect(requestClose).toHaveBeenCalled();
	});

	it('Should render settings modal component with props', () => {
		const setStateMock = jest.fn();
		useSelectorMock.mockImplementation(() => [0, 0, 0]);
		useContextMock.mockReturnValue(jest.fn());
		useStateMock.mockReturnValue([true, setStateMock]);
		useDispatchMock.mockReturnValue(jest.fn());
		const wrapper = shallow(<Dashboard />);
		expect(wrapper.find(SettingsModal).prop('onRequestClose')).toEqual(
			expect.any(Function)
		);
	});

	it('Should invoke settings modal on close correctly', () => {
		const setStateMock = jest.fn();
		useSelectorMock.mockImplementation(() => [0, 0, 0]);
		useContextMock.mockReturnValue(jest.fn());
		useStateMock.mockReturnValue([true, setStateMock]);
		useDispatchMock.mockReturnValue(jest.fn());
		const wrapper = shallow(<Dashboard />);
		wrapper.find(SettingsModal).prop('onRequestClose')();
		expect(setStateMock).toHaveBeenCalledWith(false);
	});

	it('should choose correct theme', () => {
		useContextMock.mockReturnValue(false);
		const setStateMock = jest.fn();
		useSelectorMock.mockImplementation(() => [0, 0, 0]);
		useStateMock.mockReturnValue([true, setStateMock]);
		useDispatchMock.mockReturnValue(jest.fn());
		const wrapper = shallow(<Dashboard />);
		expect(wrapper.find('div.Dashboard').get(0).props.style).toHaveProperty(
			'color',
			'black'
		);
	});

	it('Should return correct algorithms length', () => {
		const setStateMock = jest.fn();
		useSelectorMock.mockImplementation(() => [0, 0, 0]);
		useContextMock.mockReturnValue(jest.fn());
		useStateMock.mockReturnValue([true, setStateMock]);
		useDispatchMock.mockReturnValue(jest.fn());
		const wrapper = shallow(<Dashboard />);
		expect(wrapper.find(Card).at(0).prop('title')).toHaveLength(3);
	});
});
