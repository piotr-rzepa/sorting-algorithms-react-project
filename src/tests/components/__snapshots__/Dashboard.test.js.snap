// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> Should render component correctly 1`] = `
<div
  className="Dashboard"
  style={
    Object {
      "backgroundColor": "#202020",
      "color": "white",
    }
  }
>
  <SettingsModal
    contentLabel="Example Modal"
    isOpen={true}
    onRequestClose={[Function]}
  />
  <div
    className="overview"
  >
    <div
      className="overview-title"
    >
      <h1>
        Dashboard
      </h1>
      <p>
        Overview of sorting algorithms
      </p>
    </div>
    <button
      className="data-options-button"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#949496",
        }
      }
      type="button"
    >
      <ForwardRef(SvgEdit)
        height="24px"
        width="24px"
      />
      Manage data options
    </button>
  </div>
  <div
    className="small-cards"
  >
    <Card
      background={
        Object {
          "backgroundColor": "rgba(113,59,219,0.05)",
        }
      }
      icon="group.svg"
      subtitle="Algorithms"
      title={
        Array [
          0,
          0,
          0,
        ]
      }
    />
    <Card
      background={
        Object {
          "backgroundColor": "rgba(76,184,255,0.07)",
        }
      }
      icon="clock.svg"
      subtitle="Overall average sorting time"
      title="0:00:00ms"
    />
    <Card
      background={
        Object {
          "backgroundColor": "rgba(255,76,97,0.05)",
        }
      }
      icon="decreasing.svg"
      subtitle="Worst average sorting time"
      title={0}
    />
    <Card
      background={
        Object {
          "backgroundColor": "rgba(255,184,0,0.07)",
        }
      }
      icon="trophy.svg"
      subtitle="Best average sorting time"
      title={0}
    />
  </div>
  <div
    className="charts"
  >
    <ChartBar
      height={609}
      width={820}
    />
    <AlgoList />
    <CardChart
      height={609}
      width={387}
    />
  </div>
  <p
    style={
      Object {
        "padding": "2px 0",
        "textAlign": "right",
      }
    }
  >
    Made by Chili Labs & Piotr
  </p>
</div>
`;
